src/main/java/picard/illumina/ClusterDataToSamConverter.java

```diff
@@ -206,13 +237,24 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
         final String readName = readNameEncoder.generateReadName(cluster, null); // Use null here to prevent /1 or /2 suffixes on read name.
 
         // Get and transform the unmatched barcode, if any, to store with the reads
-        String unmatchedBarcode = null;
-        if (hasSampleBarcode && cluster.getMatchedBarcode() == null) {
-            final byte[][] barcode = new byte[sampleBarcodeIndices.length][];
-            for (int i = 0; i < sampleBarcodeIndices.length; i++) {
-                barcode[i] = cluster.getRead(sampleBarcodeIndices[i]).getBases();
+        final String unmatchedBarcode;
```

Comment:
There is a lot of branching here just to ensure that local variable are final.  I think it would be a lot less complicated if they weren't.

```       // Get and transform the unmatched barcode, if any, to store with the reads
        String unmatchedBarcode = null;
        String barcodeQuality = null;
        if (hasSampleBarcode && (this.barcodePopulationStrategy == PopulateBarcode.ALWAYS ||
                    this.barcodePopulationStrategy == PopulateBarcode.ORPHANS_ONLY && 
                            cluster.getMatchedBarcode() == null ||
                    this.barcodePopulationStrategy == PopulateBarcode.INEXACT_MATCH && 
                            !getUnmatchedBarcode(cluster, false).equals(cluster.getMatchedBarcode()))) {
                unmatchedBarcode = getUnmatchedBarcode(cluster, true); 
        }

        if (unmatchedBarcode != null && includeQualitiesWithBarcode){
            barcodeQuality = getBarcodeQuality(cluster);
        }
 ```

Reply:
done